src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs

```diff
@@ -182,6 +185,66 @@ public void GarbageCollect(Func<string, IWorkloadResolver> getResolverForWorkloa
             }
         }
 
+        public string InstallWorkloadSet(string path)
+        {
+            return ModifyWorkloadSet(path, InstallAction.Install);
+        }
+
+        public void RollBackWorkloadSetInstallation()
+        {
+            if (_newWorkloadSetPath is not null)
+            {
+                ModifyWorkloadSet(_newWorkloadSetPath, InstallAction.Uninstall);
+                if (_previousWorkloadSetVersion is not null)
+                {
+                    File.WriteAllText(Path.Combine(_newWorkloadSetPath, "version.txt"), _previousWorkloadSetVersion);
+                    ModifyWorkloadSet(_newWorkloadSetPath, InstallAction.Repair);
+                }
+            }
+        }
+
+        private string ModifyWorkloadSet(string path, InstallAction requestedAction)
+        {
+            ReportPendingReboot();
+
+            // Rolling back a manifest update after a successful install is essentially a downgrade, which is blocked so we have to
+            // treat it as a special case and is different from the install failing and rolling that back, though depending where the install
+            // failed, it may have removed the old product already.
+            // Resolve the package ID for the manifest payload package
```

Comment:
This isn't accurate anymore, as manifests (and workload sets) are always side-by-side, so we won't be upgrading or downgrading them.  https://github.com/dotnet/sdk/pull/38143 simplifies (ie removes) a lot of this logic.